llvm/lib/Transforms/Scalar/Scalarizer.cpp

```diff
@@ -667,14 +686,26 @@ bool ScalarizerVisitor::splitBinary(Instruction &I, const Splitter &Split) {
 bool ScalarizerVisitor::isTriviallyScalarizable(Intrinsic::ID ID) {
   if (isTriviallyVectorizable(ID))
     return true;
+  // TODO: Move frexp to isTriviallyVectorizable.
+  // https://github.com/llvm/llvm-project/issues/112408
+  switch (ID) {
+  case Intrinsic::frexp:
+    return true;
+  }
   return Intrinsic::isTargetIntrinsic(ID) &&
          TTI->isTargetIntrinsicTriviallyScalarizable(ID);
 }
 
 /// If a call to a vector typed intrinsic function, split into a scalar call per
 /// element if possible for the intrinsic.
 bool ScalarizerVisitor::splitCall(CallInst &CI) {
-  std::optional<VectorSplit> VS = getVectorSplit(CI.getType());
+  Type *CallType = CI.getType();
+  bool AreAllMatchingVectors = isStructOfMatchingFixedVectors(CallType);
+  std::optional<VectorSplit> VS;
+  if (AreAllMatchingVectors)
+    VS = getVectorSplit(CallType->getContainedType(0));
```

Comment:
There seems to be a subtle potential issue.  `VS` is presumed to be the return vector type for the function.  Later, we use `if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))` to determine whether the overload type is based on the return type, and will use this contained type.  This cannot differentiate between the elements of the struct.

Perhaps that's the intended design for this, but if so it should be made explicit, like:
> The overload type may only be dependent on the first member of the returned structure.  In other words, the other members of that returned structure can not contribute to the overload type of the intrinsic.

I think there may be another subtle problem with the component type of the other vector(s) in the returned structure.  What types must they use?  Should they match the overload type, or can they be different?  Perhaps that's a problem for earlier builtin type checking in clang, and not for the code that transforms the intrinsics, but it does seem like an issue that these other outputs are basically not addressable with the current trivially vectorizable scheme.

Reply:
I think you are confusing a few things.  First the overload  type is not based on the return type. The `isVectorIntrinsicWithOverloadTypeAtArg` has a pass through behavior of  `OpdIdx == -1`  All return indices will be true. Second We build up the return in lines 733 to 748.  `lvm.frexp.v2f64.v2i32` becomes two `lvm.frexp.f64.i32`.

Third the behavior of `isVectorIntrinsicWithOverloadTypeAtArg` isn't relevant for struct return types. The scope of this change is not intended to impact any argument overload cases.

Reply:
`isVectorIntrinsicWithOverloadTypeAtArg` defaults to returning true for `-1` because it's the common case for intrinsic overloads to be based on the return type when it matches the other arguments, but the exceptions are informative. Some of these return true for indices -1 and 0, which means the intrinsic has two overload types, one based on the return type, and one based on the first argument.  Another (`is_fpclass`) returns true only for index 0, which means the overload is based on the first argument, not the return type.

This calls `isVectorIntrinsicWithOverloadTypeAtArg` with `-1` to determine whether the first return type in the struct is used in the overload, then (see my reply to the other comment), we just assume any others that don't match the previous type in that struct are also added to the overload.  I don't think this is correct for the general case.  I think `isVectorIntrinsicWithOverloadTypeAtArg` should be augmented to support individually addressing the fields of the returned struct.  One way would be to extend into the negative space for indices: -1, -2, -3 would refer to three fields in a returned struct, in that order.

Reply:
I think that limiting our handling to cases where all of the vector types nested in the struct type of the return type match is pretty reasonable. Adding handling for cases where we have overloads in some subset of the return types seems like it would be quite complicated and I can't really come up with a realistic scenario where it might happen. The fact that the only intrinsic that exists today that has special treatment of the return type at all is `is_fpclass` is probably evidence that this is rare.

I do think a comment and/or an assert if we find that the vector types of the returned struct type mismatch is appropriate, but let's not go overboard and implement a fully generic solution for something that we only have theoretical evidence for being a problem.

Reply:
> I think that limiting our handling to cases where all of the vector types nested in the struct type of the return type match is pretty reasonable.

As Farzon points out, the `frexp` example has mismatching vector types in the return struct.  My initial assumption about how this was going to work was incorrect, so this thread may have started out with a bit of confusion.

For `frexp`, each of these are involved in the overload type for the intrinsic, so each must be captured.  The way it works in this PR now is that it captures the first field only if `isVectorIntrinsicWithOverloadTypeAtArg` called with `-1` returns true, then it captures each field after that if it differs from the prior field.  This second part sounds basically like a heuristic, not an explicit property describing a specific intrinsic.

That seems fragile, and it would be easy to break if some intrinsic required something different (even if you think that will never come up).  It's certainly possible, though likely rare, that one of the returned vectors does not contribute to the overload type, even if it differs from the other type.  Other possible scenarios couldn't be handled by this as well.

The compromise Farzon came to, after we met earlier, was to add a different query function for struct return argument fields.  At least it makes things explicit, though it means yet another function needs to be plumbed through TTI as well.

Reply:
If I'm being honest  intrinsic behaviors determined by `isVectorIntrinsicWithOverloadTypeAtArg`,  `isVectorIntrinsicWithScalarOpAtArg` and `isTriviallyVectorizable` Is kind of awful.  Adding a new one of these to handle struct return types isn't ideal. For one it means every new intrinsic with struct returns has to get added to some obscure function to get properly scalarized and in the future vectorized. Maybe thats better because it makes scalarization and in the future vectorization more explicit? Personally not a fan.

I'm also not thrilled about having to add a new TTI  api. Its a bunch of boiler plate to express something very simple. Especially since the split double case doesn't need it because the returned struct of vectors are all the same type. 

Second, the fragility IMO is overstated. The code as it is captures the pattern that exist in the current set of llvm intrinsics that represent structs by their component parts.

Third i don't agree with this being a heuristic. That implies the behavior is  best guessing at what the next overload type should be when we are being very determinstic here.

Maybe should talk about  this a bit more. If we still want to do the compromise one area to discuss is  how we can do this without a TTI API for now. I'm thinking maybe the `isStructOfMatchingFixedVectors` api could also check element type and if they are all the same we  just use the existing code path of:
```cpp
 if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))
    Tys.push_back(VS->SplitTy);
```
Which is the path splitdouble is currently taking.

Reply:
I too have a distaste for this particular property lookup pattern, as I've expressed before.  I think these should be properties defined in the intrinsic tablegen file.  Perhaps an intrinsic (vectorization) property struct/object could be defined from that and retrieved (through TTI as necessary), eliminating the need for plumbing attribute properties one by one.

> Third i don't agree with this being a heuristic.

The code was guessing whether a type is added to the overload types based on whether it was different from the previous type.  That's the heuristic.  If the types matched, that doesn't mean that the type is not used in the overload (in the general sense).